File: design-patterns/ts/state/order-status/order-context.ts
// order-context.ts

import OrderCheckState from "./order-check-state";
import OrderDeliverState from "./order-deliver-state";
import OrderInProgressState from "./order-in-progress-state";
import OrderReceiveState from "./order-receive-state";
import OrderState from "./order-state";

class OrderContext {
    private state: OrderState | null = null;
    private orderCheckState: OrderState;
    private orderInProgressState: OrderState;
    private orderDeliverState: OrderState;
    private orderReceiveState: OrderState;

    constructor() {
        this.orderCheckState = new OrderCheckState(this);
        this.orderInProgressState = new OrderInProgressState(this);
        this.orderDeliverState = new OrderDeliverState(this);
        this.orderReceiveState = new OrderReceiveState(this);

        // Set the placed state as default
        this.state = this.orderCheckState;
    }

    setState(state: OrderState | null): void {
        this.state = state;
    }

    getState(): OrderState | null {
        return this.state;
    }

    getOrderCheckState(): OrderState {
        return this.orderCheckState;
    }

    getOrderInProgressState(): OrderState {
        return this.orderInProgressState;
    }

    getOrderDeliverState(): OrderState {
        return this.orderDeliverState;
    }

    getOrderReceiveState(): OrderState {
        return this.orderReceiveState;
    }

    runNextProcess(): void {
        if (this.state != null) {
            this.state.process();
        } else {
            console.log("Order processing complete");
        }
    }
}

export default OrderContext;

File: design-patterns/go/state/order_state/order_context.go
// order_context.go

package main

import "fmt"

type OrderContext struct {
	state IOrderState

	orderCheckState      IOrderState
	orderInProgressState IOrderState
	orderDeliverState    IOrderState
	orderReceiveState    IOrderState
}

func NewOrderContext() (orderContext *OrderContext) {
	orderContext = &OrderContext{}
	orderContext.orderCheckState = NewOrderCheckState(orderContext)
	orderContext.orderInProgressState = NewOrderInProgressState(orderContext)
	orderContext.orderDeliverState = NewOrderDeliverState(orderContext)
	orderContext.orderReceiveState = NewOrderReceiveState(orderContext)
	orderContext.state = orderContext.orderCheckState
	return
}

func (orderContext *OrderContext) GetOrderCheckState() IOrderState {
	return orderContext.orderCheckState
}

func (orderContext *OrderContext) GetOrderInProgressState() IOrderState {
	return orderContext.orderInProgressState
}

func (orderContext *OrderContext) GetOrderDeliverState() IOrderState {
	return orderContext.orderDeliverState
}

func (orderContext *OrderContext) GetOrderReceiveState() IOrderState {
	return orderContext.orderReceiveState
}

func (orderContext *OrderContext) GetState() IOrderState {
	return orderContext.state
}

func (orderContext *OrderContext) SetState(state IOrderState) {
	orderContext.state = state
}

func (orderContext *OrderContext) RunNextProcess() {
	if orderContext.state != nil {
		orderContext.state.Process()
	} else {
		fmt.Println("Order processing complete")
	}
}
